language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -347,20 +347,24 @@ genDefDataType curPkgId conName mod tpls def =
                      ["daml.STATIC_IMPLEMENTS_SERIALIZABLE_CHECK<" <> conName <> ">(" <> conName <> ")"]
             in ((typeDesc, serDesc), Set.unions $ map (Set.setOf typeModuleRef . snd) bs)
         DataEnum enumCons ->
-          let
-            typeDesc =
-                [ "export enum " <> conName <> " {"] ++
-                [ "  " <> cons <> " = " <> "\'" <> cons <> "\'" <> ","
-                | VariantConName cons <- enumCons] ++
-                [ "}"
-                , "daml.STATIC_IMPLEMENTS_SERIALIZABLE_CHECK<" <> conName <> ">(" <> conName <> ")"
-                ]
-            serDesc =
-                ["() => jtv.oneOf<" <> conName <> ">" <> "("] ++
-                ["  jtv.constant(" <> conName <> "." <> cons <> ")," | VariantConName cons <- enumCons] ++
-                [");"]
+          let cs = map unVariantConName enumCons
+              typeDesc =
+                let td = ["  " <> "'" <> cons <> "'" | cons <- cs]
+                    td' = reverse td
+                in [""] <> reverse (head td' : map (<> " | ") (tail td'))
+              -- The complete definition of the companion object.
+              serDesc =
+                [ "export const " <> conName <> ": daml.Serializable<" <> conName <> "> & {"] <>
+                [ "  readonly " <> cons <> ": " <> conName <> ";" | cons <- cs ] ++
+                ["} = {"] ++
+                ["  " <> cons <> ": '" <> cons <> "'," | cons <- cs] ++
```

Comment:
One thing that has come up in the past I believe is the need to enumerate enum values (e.g., let’s say you have a dropdown or something like that). If my understanding and my google searches are correct this is easy to do with an enum but not with a union. However, given that we have the companion object, we can probably add the enumeration on that, e.g. `Color.keys = ["Red", …]`. I would suggest to leave this out of this PR and do some research if there is a common pattern for this in typescript that we can steal.

Reply:
Yes, we should definitely do something like this. If it wasn't for the `decoder` method, we could use `Object.keys(Color)`. But it's there anyway. So, having `Color.keys` would be great.

Reply:
https://github.com/digital-asset/daml/pull/5044